system/src/system_power_manager.cpp

```diff
@@ -552,52 +526,94 @@ void PowerManager::handleStateChange(battery_state_t from, battery_state_t to, b
+      float vCell = fuel.getVCell();
+      LOG_DEBUG(TRACE, "VCell: %dmV", (uint16_t)(vCell * 1000));
+      if (g_batteryState != BATTERY_STATE_DISCONNECTED) {
+        uint8_t status = power.getSystemStatus();
+        if (vCell < BATTERY_CONNECTED_VCELL_THRESHOLD && !(status & 0x08)/* MUST NOT in DPM mode, in case of sudden sink in suplement mode*/) {
+          // If battery falls below 2.0v, it's probably detached
+          battery_state_t batteryState = BATTERY_STATE_DISCONNECTED;
+          handleStateChange(g_batteryState, batteryState);
+        }
+      } else {
+        if (vCell > BATTERY_CONNECTED_VCELL_THRESHOLD) {
+          vcellDebounceCount_++;
+          if (vcellDebounceCount_ >= BATTERY_VCELL_DEBOUNCE_COUNT) {
+            battery_state_t batteryState = BATTERY_STATE_NOT_CHARGING;
+            handleStateChange(g_batteryState, batteryState);
+          }
         } else {
-          PMIC power;
-          power.setRechargeThreshold(PMIC_FAULT_RECHARGE_THRESHOLD);
-          possibleFaultCounter_ = 0;
-          faultSecondaryCounter_ = 1;
-          possibleFaultTimestamp_ = millis();
+          vcellDebounceCount_ = 0;
         }
       }
     }
-  }
-  return to;
-}
-
-void PowerManager::handlePossibleFaultLoop() {
-  if (faultSecondaryCounter_ == 1 && (millis() - possibleFaultTimestamp_ > DEFAULT_FAULT_WINDOW)) {
-      PMIC power;
-      power.setRechargeThreshold(PMIC_NORMAL_RECHARGE_THRESHOLD);
-      faultSecondaryCounter_ = 0;
-      faultSuppressed_ = millis();
+  } else {
+    // Charging is enabled
+    battery_state_t batteryState;
+    uint8_t status;
+    PMIC power(true);
+    FuelGauge fuel(true);
+    if (chargedTimeStamp_ != 0 && millis() - chargedTimeStamp_ >= BATTERY_CHARED_POSTPONE_PERIOD) {
+      chargedTimeStamp_ = 0;
+      status = power.getSystemStatus();
+      // When we get here, a charged event is raised previously followed by disabling charging,
+      // so the VBAT is not floating and the VSYS_STAT can be used to deduce the battery state.
+      // If battery is actually charged, then the battery voltage should above VSYS_MIN (3.5v)
+      // The VSYS_STAT bit should be cleared, otherwise, the battery is disconnected
+      if (status & 0x01) {
+        batteryState = BATTERY_STATE_DISCONNECTED;
+      } else {
+        // The battery may be attached during the debounce period.
+        // We need to re-check the status if device is under charging actually.
+        power.enableCharging();
+        power.enableSafetyTimer();
+        const uint8_t chrg_stat = (power.getSystemStatus() >> 4) & 0b11;
+        if (chrg_stat == 0b01 || chrg_stat == 0b10) {
+          batteryState = BATTERY_STATE_CHARGING;
+        } else {
+          batteryState = BATTERY_STATE_CHARGED;
+        }
+      }
+      handleStateChange(g_batteryState, batteryState);
+    } else if (g_batteryState == BATTERY_STATE_DISCONNECTED && millis() - disconnectedTimeStamp_ >= BATTERY_STATE_CHECK_PERIOD) {
```

Comment:
Given https://github.com/particle-iot/device-os/pull/2272/files#diff-1e4e4a03c336c7a6fe4cbb1114a4ffdd231581cf432dfa9c20587636679b2febR579, we should increase `BATTERY_STATE_CHECK_PERIOD` from 2 seconds to prevent PMIC from triggering the interrupt line too often.

Reply:
Since we've disabled charging once we got a charged event, so it won't trigger the interrupt too often. What is the period you would suggest so that it neither overload the CPU, nor take too long to update the battery state?

Reply:
Can you point me to where we are disabling charging? I don't see a call to `disableCharging()` anywhere.

Reply:
In the `handleUpdate` function where we check the charging state.

Reply:
That's a persistent setting (introduced in 3.0) that allows the application to enable or disable charging under specific conditions. Normally charging is enabled.

Reply:
This is also the code block that handles the case where charging is configured to be enabled.

Reply:
I was mentioning this: https://github.com/particle-iot/device-os/blob/563c3a5120ccb92ab17b1a7798ab56c69755cd1a/system/src/system_power_manager.cpp#L222

Reply:
Okay, it seems I misread that check. So, right now what we do is once we get charged event from PMIC and charging is enabled in power manager we are going to disable charging? I don't think we should be controlling that manually, as with disproportionate power supply vs full device current draw we may very well have to drop out of that charged state very soon and I think we should rely on the PMIC for that. The only problem is with fluctuations around recharge-threshold, but these are solvable with changing it temporarily to suppress.

This also becomes a problem for sleepy devices where manually disabling charging like this may lead to a complete discharge if we go into sleep after that.